smqtk_indexing/utils/metrics.py

```diff
@@ -97,19 +98,24 @@ def cosine_similarity(i: np.ndarray, j: np.ndarray) -> float:
     :param i: Vector i
     :param j: Vector j
 
-    :return: Float similarity.
+    :return: Float or array of cosine similarity.
     """
-    # numpy.linalg.norm is Frobenius norm (vector magnitude)
-    # return numpy.dot(i, j) / (numpy.linalg.norm(i) * numpy.linalg.norm(j))
+    if i.ndim == 1:
+        i = i.reshape(1, -1)
+    if j.ndim == 1:
+        j = j.reshape(1, -1)
 
-    # speed optimization, numpy.linalg.norm can be a bottleneck
-    # TODO(john.moeller): This can be made array-friendly
-    #   np.multiply(i, j).sum(axis) / (np.linalg.norm(i, i_axis) *
-    #       np.linalg.norm(j, j_axis))
-    return np.dot(i, j) / (np.sqrt(i.dot(i)) * np.sqrt(j.dot(j)))
+    # Cosine similarity function returns matrix in shape of (i_N_samples, j_N_samples),
+    cosine_s = 1 - cdist(i, j, metric='cosine')[0]
 
+    # Return a float if both i and j only have one sample
+    if cosine_s.size == 1:
+        return cosine_s[0]
+    else:
+        return cosine_s
```

Comment:
We should add some tests for this metric: I don't think `cosine_s[0]` is enough extraction:
```python
In [1]: import numpy as np

In [2]: from scipy.spatial.distance import cdist

In [3]: cdist( [[0, 1]], [[0,1]])
Out[3]: array([[0.]])

In [4]: cdist( [[0, 1]], [[0,1]])[0]  # outputs an array of size 1 and shape (1,)
Out[4]: array([0.])

In [5]: cdist( [[0, 1]], [[0,1]])[0,0]  # outputs a scalar np.float64
Out[5]: 0.0

```

Reply:
Oops, nvm, missed the `[0]` indexing on L109.